pytorch_lightning/trainer/training_loop.py

```diff
@@ -529,44 +528,38 @@ def run_training_epoch(self):
 
             self.total_batch_idx += 1
 
-            max_steps_reached = (
-                self.max_steps is not None and self.max_steps <= self.global_step + 1
-                and self._accumulated_batches_reached()
-            )
-            if max_steps_reached or self.trainer.should_stop or self._num_training_batches_reached(is_last_batch):
-                break
-
             # progress global step according to grads progress
             self.increment_accumulated_grad_global_step()
 
+            max_steps_reached = (self.max_steps is not None and self.max_steps <= self.global_step)
+            if max_steps_reached or self.trainer.should_stop or self._num_training_batches_reached(is_last_batch):
+                break
+
         if batch_idx is None:
             # dataloader/iterator did not produce a batch
             return
 
         # handle epoch_output on epoch end
         self.on_train_epoch_end(epoch_output)
 
+        # the global step is manually decreased here due to backwards compatibility with existing loggers
+        # as they expect that the same step is used when logging epoch end metrics even when the batch loop has
+        # finished. this means the attribute does not exactly track the number of optimizer steps applied.
+        # TODO(@carmocca): deprecate and rename so users don't get confused
+        self.global_step -= 1
         # log epoch metrics
         self.trainer.logger_connector.log_train_epoch_end_metrics(epoch_output)
+        self.global_step += 1
 
-        should_check_val = self._should_check_val_fx(batch_idx, is_last_batch, on_epoch=True)
-        should_skip_eval = self.trainer.evaluation_loop.should_skip_evaluation(self.trainer.num_val_batches)
-        should_train_only = self.trainer.disable_validation or should_skip_eval
-
-        # update epoch level lr_schedulers if no val loop outside train loop is triggered
-        if not should_check_val or should_train_only:
-            self.update_lr_schedulers('epoch')
+        self.update_lr_schedulers('epoch')
 
-        if should_train_only:
+        did_train_only = self.trainer.disable_validation or self.trainer.evaluation_loop.should_skip_evaluation(
+            self.trainer.num_val_batches
+        )
+        if did_train_only:
+            self.global_step -= 1
             self.check_checkpoint_callback(True)
-
-        if should_check_val:
-            self.trainer.validating = True
-            self.trainer._run_evaluation(on_epoch=True)
-            self.trainer.training = True
-
-        if batch_output.signal != -1:
-            self.increment_accumulated_grad_global_step()
+            self.global_step += 1
```

Comment:
nit: could we add a TODO that this should be removed in favor of pushing this logic directly to the checkpoint callback?

Reply:
With "this" do you mean the `did_train_only` check, the `global_step` change, or both?

Reply:
the whole `if did_train_only` force-calling the checkpoint callback

Reply:
Yes

Reply:
Answering @ananthsub 

> the whole if did_train_only force-calling the checkpoint callback

I'm not sure we want to move it inside the callback as it is logic tied to the loop structure.

Reply:
> Yes

Then I think this is a great clean up.

I think what @ananthsub is saying is let the callback system call ModelCheckpoint. I think we should aim for that.

Reply:
Discussed offline with Adrian. Not going to add the TODO but will try to resolve this asap.